### PR TITLE
Retrigger existing jobs N times (when possible) instead of triggering N arbitrary jobs

### PR DESCRIPTION
**Pull request by adusca**

### CODE REVIEW

**Comment by adusca** [comment]
Output (with --dry-run):

mozilla_ci_tools[retrigger-existing]$ python mozci/scripts/trigger.py -b "Ubuntu VM 12.04 x64 try debug test cppunit" -r 062dfff3dbbd --dry-run --times 10
05/27/2015 10:28:33 INFO:    We want to have 10 job(s) of Ubuntu VM 12.04 x64 try debug test cppunit on revisions ['062dfff3dbbd']
05/27/2015 10:28:33 INFO:  
05/27/2015 10:28:33 INFO:    === 062dfff3dbbd ===
05/27/2015 10:28:34 INFO:    We want to have 10 job(s) of Ubuntu VM 12.04 x64 try debug test cppunit on revision 062dfff3dbbd
builds-4hr.js: [>>>>>>>>>>>>>>>>>>>>>>] Elapsed Time: 0:00:03 714.15 kB/s 2.0MB
05/27/2015 10:28:40 INFO:    We have found 1 potential job(s) matching 'Ubuntu VM 12.04 x64 try debug test cppunit' on 062dfff3dbbd. We need to trigger more.
05/27/2015 10:28:40 INFO:    We would make a POST request to https://secure.pub.build.mozilla.org/buildapi/self-serve/try/request with the payload: {'count': 9, 'request_id': 70746749}
05/27/2015 10:28:40 INFO:    https://treeherder.mozilla.org/#/jobs?repo=try&fromchange=062dfff3dbbd&tochange=062dfff3dbbd&filter-searchStr=Ubuntu+VM+12.04+x64+try+debug+test+cppunit

File: mozci/sources/buildapi.py [hunk 1]
```diff
@@ -67,6 +67,52 @@ def make_request(repo_name, builder, revision, files=[], dry_run=False,
     return req
 
 
+def make_retrigger_request(repo_name, request_id, dry_run=True, count=1):
```
Comment (armenzg) [inline mozci/sources/buildapi.py]: Could you please add priority as an optional value defaulting to 0?
I think this applies in here too: "Higher priority values get handled first, and the default priority for jobs is 0."

Could you please put dry_run as the last parameter?

File: mozci/sources/buildapi.py [hunk 2]
```diff
@@ -67,6 +67,52 @@ def make_request(repo_name, builder, revision, files=[], dry_run=False,
...
+    Rebuild `request_id`, which must be passed in as a POST parameter.
+    `priority` and `count` are also accepted as optional
+    parameters. `count` defaults to 1, and represents the number
+    of times this build  will be rebuilt.
+    """
+    url = '{}/{}/request'.format(HOST_ROOT, repo_name)
+    payload = {'request_id': request_id}
+
+    if count != 1:
+        payload.update({'count': count})
```
Comment (armenzg) [inline mozci/sources/buildapi.py]: We can add priority in here IIUC.

File: mozci/sources/buildapi.py [hunk 3]
```diff
@@ -67,6 +67,52 @@ def make_request(repo_name, builder, revision, files=[], dry_run=False,
...
+
+    if dry_run:
+        LOG.info('We would make a POST request to %s with the payload: %s' % (url, str(payload)))
```
Comment (armenzg) [inline mozci/sources/buildapi.py:89]: Could you add a bit of info outside of the dry_run block?
"We're going to re-trigger an existing completed job %s times."

File: mozci/sources/buildapi.py [hunk 4]
```diff
@@ -67,6 +67,52 @@ def make_request(repo_name, builder, revision, files=[], dry_run=False,
...
+def make_cancel_request(repo_name, request_id, dry_run=True):
+    """
+    Cancel a request using buildapi self-serve. Returns a request.
+
+    Builapi documentation:
+    DELETE /self-serve/{branch}/request/{request_id} Cancel the given request
+    """
+    url = '{}/{}/request/{}'.format(HOST_ROOT, repo_name, request_id)
+    if dry_run:
+        LOG.info('We would make a DELETE request to %s.' % url)
```
Comment (armenzg) [inline mozci/sources/buildapi.py:112]: Could you please add an info message outside of the dry_run block?
That way the dev can know that a job got canceled.

File: mozci/sources/buildapi.py [hunk 5]
```diff
@@ -67,6 +67,52 @@ def make_request(repo_name, builder, revision, files=[], dry_run=False,
...
+        return None
+
+    req = requests.delete(url, auth=get_credentials())
```
Comment (armenzg) [inline mozci/sources/buildapi.py:116]: I think we could add a debug message that gives the URL to inspect a deletion or a retrigger:
e.g. https://secure.pub.build.mozilla.org/buildapi/self-serve/jobs/1337465
I can't remember if this URL or some metadata is returned inside of `req`.

**Comment by armenzg** [comment]
Hi Alice,
We don't need to wait for some of the changes that I mentioned since I would not want to delay your current work.
Feel free to create follow up issues to keep track of them and deal with them later (e.g. add debug messages to the URL that represents a job cancellation)

**Comment by adusca** [comment]
I added priority as a parameter and LOG.infos after dry_run.

About the debug message with the URL: make_retrigger_request and make_cancel_request only have the request_id, so the URL we start with is https://secure.pub.build.mozilla.org/buildapi/self-serve/{branch}/request/{request_id} (e.g. https://secure.pub.build.mozilla.org/buildapi/self-serve/try/request/70772496 ) and doesn't contain very useful information.

The URL that shows cancel/retrigger is https://secure.pub.build.mozilla.org/buildapi/self-serve/jobs/{job_id} (e.g. https://secure.pub.build.mozilla.org/buildapi/self-serve/jobs/1339071 ) but I don't see a way to get the job_id we need from the URL we have without querying every job in /self-serve/jobs and see the ones that match.

In mozci that wouldn't be a problem, but I'm afraid of making pulse_actions too slow to handle all incoming requests. I have no idea how fast pulse_actions has to be. What do you think?

**Comment by armenzg** [comment]
Great analysis! Let's keep it fast. Let me look at the code, merge it and do the release.

**Pull request merged**